File: stickynote-loader/stickynote-loader-bukkit/src/main/java/org/sayandev/stickynote/loader/bukkit/StickyNoteBukkitLoader.java
package org.sayandev.stickynote.loader.bukkit;

import com.alessiodp.libby.BukkitLibraryManager;
import com.alessiodp.libby.LibraryManager;
import com.alessiodp.libby.PaperLibraryManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.sayandev.loader.common.StickyNoteLoader;
import org.sayandev.stickynote.bukkit.WrappedStickyNotePlugin;

import java.io.File;
import java.util.logging.Logger;

public class StickyNoteBukkitLoader extends StickyNoteLoader {

    public JavaPlugin javaPlugin;

    public StickyNoteBukkitLoader(JavaPlugin javaPlugin) throws NoSuchFieldException, ClassNotFoundException, IllegalAccessException {
        this(javaPlugin, false);
    }

    public StickyNoteBukkitLoader(JavaPlugin javaPlugin, boolean reloadStickyNote) throws NoSuchFieldException, ClassNotFoundException, IllegalAccessException {
        super(javaPlugin.getName());
        this.javaPlugin = javaPlugin;

        File dataFolder = javaPlugin.getDataFolder();
        Logger logger = javaPlugin.getLogger();

        LibraryManager libraryManager = getLibraryManager(javaPlugin);
        this.load(javaPlugin.getName(), dataFolder.getParentFile(), logger, libraryManager, false);
    }

    @Override
    protected void onComplete() {
        new WrappedStickyNotePlugin(javaPlugin);
    }

    private static LibraryManager getLibraryManager(JavaPlugin plugin) {
        if (plugin.getResource("paper-plugin.yml") != null) {
            plugin.getLogger().info("paper-plugin detected, using paper library loader...");
            return new PaperLibraryManager(plugin, new File(plugin.getDataFolder().getParentFile(), "stickynote").getAbsolutePath());
        } else {
            return new BukkitLibraryManager(plugin, new File(plugin.getDataFolder().getParentFile(), "stickynote").getAbsolutePath());
        }
    }
}

File: stickynote-loader/stickynote-loader-bukkit/src/main/kotlin/org/sayandev/stickynote/loader/bukkit/StickyNoteBukkitLoader.kt
package org.sayandev.stickynote.loader.bukkit

import com.alessiodp.libby.BukkitLibraryManager
import com.alessiodp.libby.Library
import org.bukkit.plugin.java.JavaPlugin
import org.sayandev.stickynote.bukkit.WrappedStickyNotePlugin

object StickyNoteBukkitLoader {

    @JvmStatic
    fun load(plugin: JavaPlugin) {
        load(plugin, false)
    }

    @JvmStatic
    fun load(plugin: JavaPlugin, usePaperLoader: Boolean) {
        val stickyNotes = Class.forName("org.sayandev.stickynote.generated.StickyNotes")
        val dependencies = stickyNotes.fields.filter { it.name.startsWith("DEPENDENCY_") }.map { it.get(null) }
        val repositories = stickyNotes.fields.filter { it.name.startsWith("REPOSITORY_") }.map { it.get(null) } as List<String>

        // TODO: utilize paper library manager
        val libraryManager = if (usePaperLoader) BukkitLibraryManager(plugin) else BukkitLibraryManager(plugin)
        libraryManager.addMavenLocal()
        libraryManager.addRepository("https://repo.sayandev.org/snapshots")
        for (repository in repositories) {
            libraryManager.addRepository(repository.replace("{}", "."))
        }

        for (dependency in dependencies) {
            val group = (dependency::class.java.getMethod("getGroup").invoke(dependency) as String).replace(".", "{}")
            libraryManager.loadLibrary(
                Library.builder()
                    .groupId(group)
                    .artifactId(dependency::class.java.getMethod("getName").invoke(dependency) as String)
                    .version(dependency::class.java.getMethod("getVersion").invoke(dependency) as String)
                    /*.apply {
                        if (relocationFrom?.isNotEmpty() == true && relocationTo?.isNotEmpty() == true) {
                            this.relocate(group, relocationTo)
                        }
                    }*/
                    .build()
            )
        }

        WrappedStickyNotePlugin(plugin)
    }

}

